### DISCUSSION TITLE
Dropdowns should be a component in @nativescript/plugin

### DISCUSSION THREAD

**Discussion by AnthonyLenglet** [NativeScript/rfcs#23] [1. Ideas & Discussions]
As mentioned in the title, it feels weird that one the most popular Nativescript plugin is only supporting Nativescript pre 6.0, with the updated version being locked behind a subscription service
https://github.com/PeterStaev/NativeScript-Drop-Down

The other frameworks have pretty official looking and well maintained plugins for this
https://github.com/react-native-picker/picker
ionic also has an official component
https://ionicframework.com/docs/api/picker

overall I feel like the subscription service that Nativescript has, while good for open source dev, is hindering its adoption, when every other framework has a much bigger community and isn't even doing something like this on the scale NS is https://npm.proplugins.org/
(of cours thats another discussion, but its definitely part of the issue, because the community around NS is so small, people are alone working on those plugins and have to resort to systems like these, which in turn stops people from joining the community)

**Comment by rigor789** [comment]
Just to clarify - ProPlugins is not an official marketplace for NativeScript and it's not endorsed by NativeScript in any way.

👍 on including an official picker (can be built with ListPicker or ActionDialog + a bit of custom layout).

**Comment by NathanaelA** [comment]
First I am one of the founders of nStudio, so we currently maintain the NS Framework.  
Second, in defense of ProPlugins; I am also the founder of ProPlugins.     Please note; nStudio does NOT run, own or have anything to do with ProPlugins, they are two totally separate companies, ProPlugins & all the nativescript.rocks sites are owned by Master Technology (one of my other companies) and both are something I started **LONG** before nStudio took over NativeScript.      You are not the only person who doesn't like ProPlugins, in fact some of the nStudio team does not like ProPlugins for some of the same reasons you state, and we have had some interesting internal discussions about ProPlugins and its place moving forward (nothing resolved yet) since nStudio took over NativeScript maintenance last year.    However, a decent chunk of the community has thanked me personally for starting it, which is amazing to me that I have seen more praise than hate for the idea/implementation, since most people statistically are likely to complain...

So either people love me or hate me for starting it.  But I figured I should give you the disclaimer so you know from where I speak, as I wear both hats...    :grinning:

---
Lets look at the facts: (Some of them you alluded to)

1. Open source has a major issue with funding, some projects get a lot of funding, most don't actually get enough to survive.

2. Open source has a major issue with people helping work on it, some projects do get a lot of people helping.  Most  don't see any help (which is especially true of virtually all the plugins).   Of the 40+ plugins in ProPlugins, all of them that started as open source are still left in the exact same open source repos at the point of forking to allow the community to fully maintain them after that point.  Think of ProPlugins as the paid toll road that leads to your destination much much quicker, the community is often unpaved road that takes a lot longer and might require some detours, but you can still get to your destination.

3.  So far, I think in the ~2 years since the starting of ProPlugins, I have personally seen  ~3 PR's from the community to fix one of these community supported plugins.   Each PR I've merged and pushed a new community version within a couple days, but still only ~3 PR's in two years for ~40 plugins....  Nobody else wants to take on the "cost" of maintenance, so the roads stay unpaved and slowly gets potholes. 

4. The authors of all these plugins also have families to feed; unpaid work is unfortunately not sustainable and ultimately ends up being harmful to everyone especially in the context of people needing to depend on the plugins for there apps and them not working as this can have severe business issues for companies. (which is the primary reason most companies do like ProPlugins, they know they can depend on the high quality of them for such a minor cost, and don't have to expend any additional resources to fix any of the potholes)

5. Proplugins is actually a very tiny part of the NS plugin community if you use one of *my* other free resources; https://plugins.nativescript.rocks -- you will see as of this moment their is > 1,500 plugins.   Of that ~40 are ProPlugins.    That means that  less than 3% are ProPlugins, almost 98% are not ProPlugins, including still a number of mine which I didn't move.   So ProPlugins is actually a very tiny amount of the plugins in total eco-system, we just have some of the best maintained plugins (discuss more **later**)

6. The most popular plugin in the eco-system, is my NS-Permissions; it is still a community plugin and I still maintain it for **free** as it is used in so many places in the entire eco-system, moving it to be exclusively pro-plugins would be a disaster for the community and as such that was the reason I didn't move it to ProPlugins.   ProPlugins was never meant to crippled the eco-system, it was meant to be an additional resource of higher quality plugins.

---
Now lets look specifically at your example:

1.    React-native-picker is open source, and the community is now fully maintaining it.     So the RN plugin is no longer a officially support plugin, as the **official** plugin was depreciated (and appears turned into this one, to be fully maintained by the community).  Think about that for a sec, the multi-billion dollar company didn't want to support it, and they would rather foist it and its support off onto the community and not pay for it.   As a side note to discuss **later**; this repo currently has 58 open issues, some appear to be crashing issues. 

2.  NS-DropDown has been open source  and the version you linked to was left to be maintained by the community (can't say its state; but the code is available for ANYONE including **you** to maintain, fork, and fix.)   So it is **exactly** like the React-Native-Picker is right now, both are community maintained.    Based on the issues in the RN plugin, it appears it is moving towards (or already is) a buggy plugin.  It really is hard to support a plugin for free and again the multi-billion dollar Facebook didn't even want to continue to support it (& several other components they also spun off to the community) either.

3.   The ProPlugins/Drop-Down version the author has maintained it (looks to be about ~30 PR's since joining ProPlugins), he currently has **1** open issue on this repo, and that issue is a end user error based on what I just read.   (Again to discuss **later**)

---
**Later**:  :grinning: 

ProPlugins currently has ~4 open issues/feature requests across all of its ~40 plugins.     Think about that, less than 1 open issue per every 10 plugins and most of those "issues" are actually just feature requests.   
Now lets look at the links you sent; 58 issues in just that single RN plugin repo, that is 58 issues for ONE single plugin.  Ionic has ~29 open issues just for the "picker" and another 109 open issues for the "select" (similar plugin) in the Ionic repo.   In summary, you do get what you pay for.     Which would you rather have, pay a tiny fee and have ~0 issues and have access to more than 40 fully supported  different plugins?    Or pay nothing and have unsupported and potentially have to spend time fixing any issues you run into, what is your time worth to you?

So most people who dislike ProPlugins they see it as a **"either/or"** choice, ProPlugins is bad because it supplants the open source...   However, in all reality, ProPlugin is a **"ALSO"**,  what do I mean by that....  

The interesting thing is, you actually have a choice in the NativeScript community.   The open source "community" maintained plugin are still 100% available, and you can spend your time fixing it and making it work for your app and so can anyone in the community and do PR's to it, nothing is stopping anyone from doing this.    This is the typical scenario in a open source project, and it is exactly the same for the current drop down plugin as it is for the other 1,500 plugins in the eco-system.   Your freedom has not been limited.   But only in the NativeScript do you **ALSO** have the choice of using a fully maintained, high quality, and well rounded plugins.      In the other communities you have no choice; you may have to "fix" it to use it, if you run into any issues.    This is what I mean by **ALSO**; as you now have a choice of using X and **ALSO** Y.

---

Now as for your actual idea; I am mostly against it (not because it is part of ProPlugins), but I am against pretty much against adding any **new** additional UI components to core.    In fact I actually would like to **remove** a few that are rarely used and/or are better maintained in the community. :grinning:     I do have to admit this one straddles that line as it could be useful in a lot of applications, which is why there is some support from me.    But as an objective whole, I feel it still falls short of the criteria to be added to the official core UI as we really don't need more code to maintain and need to simplify the core we do have to make it easier to maintain for everyone who is doing this because they love the platform as most work on NativeScript is still unpaid.   

However, if you would like to volunteer to be a the maintainer of this new UI plugin, I would be all for that and would gladly support you in it.   I would love to see more people from the community take over parts of the official framework kinda like Debian has maintainers for each package, I would love to see maintainers for each UI component.   So if you are volunteering to update it and maintain it, I can :100: get behind that!   Otherwise, my vote is no.

**Comment by AnthonyLenglet** [comment]
Hey ! lots of interesting feedback here, thanks a ton for the in-depth response ! 😄

As a quick note before I get into it, sorry if I sounded overly negative about proplugin, as I mention in the OP, it is definitely something that needs to be looked into and I would love to see more of that !
the one issue I'd really say about it is the fact that I could not find any info on the subscription rate, since it looks like you have to log in to see it, which feels like its hiding a huge price point :/

One thing that could be cool to see (and maybe was already mentioned internally), would be to merge proplugin with the nativescript marketplace, since, and I'm only talking for our (small team of 2/3 people) side here, we have never willingly gone into this marketplace to look for a plugin, we only ever got there via a search result link expecting it'd send me to npm or github.
Not only would this make it official, but it could also make people more willing to join it, seing it on the official site rather than a small side thing that you don't know the state of right away (if needed I could open a separate discussion about this one 😄)

As for another point, which sadly I don't see a real solution for (at least as of writing this), is the amount of PRs being opened, I think the biggest pain point is the knowledge needed to work on those plugins which seems really high, I know its a point that's preventing us from helping with this, since we're a angular development startup (out of like 15 projects going on "at the same time", this is the only one using being a mobile app, so learning the stuff needed to help with nativescript would only help with this one project, so in our case it isn't as worth it as a company doing even 50/50 web and mobile app)

> Now lets look at the links you sent; 58 issues in just that single RN plugin repo, that is 58 issues for ONE single plugin. Ionic has ~29 open issues just for the "picker" and another 109 open issues for the "select" (similar plugin) in the Ionic repo. In summary, you do get what you pay for.

I've actually checked for the ionic component, is it not an official one ? it seems to be in the framework repository

As for the rest, while this definitely has an impact, there is also the amount of features (in the case of the dropdown, it seems to be about the same in that aspect, its mainly in terms of OSs where the RN version has support for windows and macOS + two different styles for android (popup and dropdown)) and the install base that coincides with it (if we go by npm numbers, RN picker has 44 027, NS Dropdown has 668 (not counting the proplugin version I'm assuming)), both of those are going to end in a bigger possibility that bugs are found

> In fact I actually would like to remove a few that are rarely used and/or are better maintained in the community. 😀

OH OH I KNOW THIS ONE ! IS IT RADGAUGE ? 😄

**Comment by NathanaelA** [comment]
> As for another point, which sadly I don't see a real solution for (at least as of writing this), is the amount of PRs being opened, I think the biggest pain point is the knowledge needed to work on those plugins which seems really high, I know its a point that's preventing us from

Yes, this is a pain point; we haven't figured out a good solution for it either....  :face_with_head_bandage: 


> the one issue I'd really say about it is the fact that I could not find any info on the subscription rate, since it looks like you have to log in to see it, which feels like its hiding a huge price point :/

Interesting, I thought I had the pricing on the site before logging in.    That is my bad, I'll have to go see what happened to it or if I need to make the link easier to find...   Thanks for the feedback.  :grinning:


> One thing that could be cool to see (and maybe was already mentioned internally), would be to merge proplugin with the nativescript marketplace, since, and I'm only talking for our (small team of 2/3 people) side here, we have never willingly gone into this marketplace to look for a plugin, we only ever got there via a search result link expecting it'd send me to npm or github.

Last year I did create additional hooks from https://plugins.nativescript.rocks to the npm.proplugins.org so all plugins listed on ProPlugins should be searchable now on  the best plugin site.  :grinning:      As for the Marketplace plugin site, we are trying to keep things separate so that people don't assume that nStudio supports ProPlugins as their are people in the community (& at nStudio) that dislike it.  It is a weird line we walk on it because I started ProPlugins long before nStudio took over NativeScript's maintenance and so it has become kinda seen in some people as a conflict of interest in some ways as they don't always see that ProPlugins is a additional option...   People frequently look at the surface, and don't always realize I am so committed to NativeScript, that I've already moved some stuff that was exclusive to ProPlugins into the NS core in NS 7.  And there are two other items in ProPlugins that I also believe should be in core and I hope will make it in NS 8...  :grinning:     

> Not only would this make it official, but it could also make people more willing to join it, seing it on the official site rather than a small side thing that you don't know the state of right away (if needed I could open a separate discussion about this one smile)

I'm not sure nStudio would ever support it officially, my vision on how to improve the eco-system is a lot different than some of the other partners at nStudio, I haven't been able to sell them fully on my vision.   TLDR; I tend to think that people and companies will use a eco-system if they can find high quality plugins that match what they need to do easily.   If they spend hours wasting time finding and installing plugins that don't work or maybe barely work; they will tend to get frustrated and leave and go to a different eco-system where the plugins are more easily found.    So over the years, I've implemented pieces to make this vision a reality; https://plugins.nativescript.rocks was created to catalog plugins and allow easily filtering.,   and ProPlugins  (to maintain them) were both started to fill parts of that vision...  

> same in that aspect, its mainly in terms of OSs where the RN version has support for windows and macOS + two different styles for android (popup and dropdown)) and the install base that coincides with it (if we go by npm numbers, RN picker has 44 027, NS Dropdown has 668

Yeah, unfortunately the npm numbers wouldn't have good numbers for anything in the ProPlugins since it can't track it since all the downloads all go thru a private npm server.   The Dropdown version on the public NPM, to my knowledge, hasn't worked with NS since NS5, so it is actually very surprising that it is still getting 688 downloads a month two years later for people actually using that old of a framework version.  

> I've actually checked for the ionic component, is it not an official one ? it seems to be in the framework repository

To my knowledge this is the official version ion-picker & ion-select; all depends on what you need, which is why I gave both open issues numbers.


>> In fact I actually would like to remove a few that are rarely used and/or are better maintained in the community. grinning

> OH OH I KNOW THIS ONE ! IS IT RADGAUGE ? smile

Yeah, the RAD**** some days is a major pain -- they should be re-written to be open source; but that is a totally different task.  :grinning:    No, I'm thinking things like the bottombar; the community maintained version by Martin is actually a much much much better component than the very limited one we have in the framework, and is pretty much drop in replaceable with many additional features.   However, I think moving each piece into a separate maintainable plugin would make it easier for people to be able to pick up some work in them without being overwhelmed.
